From 2be57f27e8a8ae447c45524a3735ee02727b0345 Mon Sep 17 00:00:00 2001
From: Rosen Penev <rosenp@gmail.com>
Date: Thu, 27 Aug 2020 18:19:04 -0700
Subject: [PATCH] hwclock: use settimeofday syscall directly

musl's settimeofday implementation completely ignores the tz parameter.
While this is perfectly valid, it causes problems when wanting to set
the kernel timezone.

Added comments.

Signed-off-by: Rosen Penev <rosenp@gmail.com>
---
 util-linux/hwclock.c | 17 +++++++++++++----
 1 file changed, 13 insertions(+), 4 deletions(-)

diff --git a/util-linux/hwclock.c b/util-linux/hwclock.c
index e85bca2b2..0c4a5c9ec 100644
--- a/util-linux/hwclock.c
+++ b/util-linux/hwclock.c
@@ -34,6 +34,7 @@
 
 #include "libbb.h"
 /* After libbb.h, since it needs sys/types.h on some systems */
+#include <sys/syscall.h>
 #include <sys/utsname.h>
 #include "rtc_.h"
 
@@ -49,6 +50,10 @@
  */
 #define SHOW_HWCLOCK_DIFF 0
 
+/* musl 1.2.0+ does not have SYS_settimeofday */
+#ifndef SYS_settimeofday
+#define SYS_settimeofday SYS_settimeofday_time32
+#endif
 
 #if !SHOW_HWCLOCK_DIFF
 # define read_rtc(pp_rtcname, sys_tv, utc) read_rtc(pp_rtcname, utc)
@@ -129,8 +134,10 @@ static void to_sys_clock(const char **pp_rtcname, int utc)
 	 */
 	tz.tz_dsttime = 0;
 
-	/* glibc v2.31+ returns an error if both args are non-NULL */
-	if (settimeofday(NULL, &tz))
+	/* glibc v2.31+ returns an error if both args are non-NULL
+	 * musl completely ignores the tz parameter
+	 */
+	if (syscall(SYS_settimeofday, NULL, &tz))
 		bb_perror_msg_and_die("settimeofday");
 
 	tv.tv_sec = read_rtc(pp_rtcname, NULL, utc);
@@ -288,8 +295,10 @@ static void set_system_clock_timezone(int utc)
 	if (!utc)
 		tv.tv_sec += tz.tz_minuteswest * 60;
 
-	/* glibc v2.31+ returns an error if both args are non-NULL */
-	if (settimeofday(NULL, &tz))
+	/* glibc v2.31+ returns an error if both args are non-NULL
+	 * musl completely ignores the tz parameter
+	 */
+	if (syscall(SYS_settimeofday, NULL, &tz))
 		bb_perror_msg_and_die("settimeofday");
 	if (settimeofday(&tv, NULL))
 		bb_perror_msg_and_die("settimeofday");

